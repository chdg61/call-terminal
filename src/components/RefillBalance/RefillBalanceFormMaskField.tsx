import classnames from 'classnames';
import * as React from 'react';
import MaskedInput from 'react-text-mask';

interface Props {
    name: string;
    label: string;
    mask: any;
    value: string;
    error: string;
    touched: boolean;
    onChange?: React.ChangeEventHandler<HTMLInputElement>;
    onBlur?: React.FocusEventHandler<HTMLInputElement>;
    placeholder?: string;
}


export class RefillBalanceFormMaskField extends React.Component<Props> {

    constructor(props: Readonly<Props>) {
        super(props);

        this.onBlur = this.onBlur.bind(this);
        this.onChange = this.onChange.bind(this);
    }

    static defaultProps: Partial<Props> = {
        placeholder: '',
    }

    private onChange(e: React.ChangeEvent<HTMLInputElement>){
        if(this.props.onChange && typeof this.props.onChange === 'function'){
            this.props.onChange(e);
        }
    }

    private onBlur(e: React.FocusEvent<HTMLInputElement>){
        if(this.props.onBlur && typeof this.props.onBlur === 'function'){
            this.props.onBlur(e);
        }
    }

    render(): React.ReactElement<any, string | React.JSXElementConstructor<any>> | string | number | {} | React.ReactNodeArray | React.ReactPortal | boolean | null | undefined {
        const {
            name,
            label,
            mask,
            touched,
            error,
            value,
            placeholder
        } = this.props;

        return (
            <React.Fragment>
                <label htmlFor={name}>
                    {label}
                </label>
                <MaskedInput
                    mask={mask}
                    className={classnames('form-control', {
                        'is-invalid': error && touched,
                    })}
                    value={value}
                    id={name}
                    placeholder={placeholder}
                    onBlur={this.onBlur}
                    onChange={this.onChange}
                    name={name}
                    guide
                />
                {touched && !!error && (
                    <div className="invalid-feedback">
                        {error}
                    </div>
                )}
            </React.Fragment>
        );
    }
}
