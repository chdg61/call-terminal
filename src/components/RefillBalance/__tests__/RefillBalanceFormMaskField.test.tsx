import { mount, shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import MaskedInput from 'react-text-mask';
import { RefillBalanceFormMaskField } from '../RefillBalanceFormMaskField';


describe('RefillBalanceFormMaskField', () => {
    let component: ShallowWrapper<RefillBalanceFormMaskField>;
    let onBlur: any;
    let onChange: any;

    let createComponent = (value: string, error: string, touched: boolean) => {
        return (
            <RefillBalanceFormMaskField
                mask={['+', '7', ' ', '(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/]}
                label="Номер телефона"
                name="phone-test"
                placeholder="Введите номер телефона"
                value={value}
                onBlur={onBlur}
                onChange={onChange}
                error={error}
                touched={touched}
            />
        );
    }

    beforeEach(() => {
        onBlur = jest.fn();
        onChange = jest.fn();
        component = shallow(createComponent('', '', false));
    })

    it('default render', () => {
        expect(component).toBeDefined();
        expect(component.html()).toMatchSnapshot();
    });

    it('change value with reformat by mask - success', (done) => {
        const component = mount(createComponent('1234567890', '', false));
        setImmediate(() => {
            component.update();
            expect(component.find(MaskedInput).find('input').instance().value).toBe('+7 (123) 456-7890');
            expect(onChange).not.toBeCalled();
            expect(component.find('.invalid-feedback').length).toBe(0);
            done();
        })
    });

    it('change value with reformat by mask - part value', (done) => {
        const component = mount(createComponent('12345', '', false));
        setImmediate(() => {
            component.update();
            expect(component.find(MaskedInput).find('input').instance().value).toBe('+7 (123) 45_-____');
            expect(onChange).not.toBeCalled();
            expect(component.find('.invalid-feedback').length).toBe(0);
            done();
        })
    });

    it('onChange value', (done) => {
        const component = mount(createComponent('12345', '', false));
        const inputElement = component.find('input');
        inputElement.simulate('change', {
            persist: () => {},
            target: {
                name: 'inputElement',
                value: '+7 (123) 45_-____'
            }
        });
        setImmediate(() => {
            component.update();
            expect(onChange).toBeCalled();
            expect(onChange.mock.calls[0][0].target).toEqual({
                name: 'inputElement',
                value: '+7 (123) 45_-____'
            });
            done();
        })
    });

    it('onBlur value', (done) => {
        const component = mount(createComponent('12345', '', false));
        const inputElement = component.find('input');
        inputElement.simulate('blur');
        setImmediate(() => {
            component.update();
            expect(onBlur).toBeCalled();
            done();
        })
    });

    it('set error', (done) => {
        const component = mount(createComponent('12345', 'set error', false));
        const inputElement = component.find('input');
        setImmediate(() => {
            component.update();
            expect(component.find('.invalid-feedback').length).toBe(0);
            expect(component.find('input').hasClass('is-invalid')).toBeFalsy();
            done();
        })
    });

    it('set touch', (done) => {
        const component = mount(createComponent('12345', '', true));
        const inputElement = component.find('input');
        setImmediate(() => {
            component.update();
            expect(component.find('.invalid-feedback').length).toBe(0);
            expect(component.find('input').hasClass('is-invalid')).toBeFalsy();
            done();
        })
    });

    it('set touch and error', (done) => {
        const component = mount(createComponent('12345', 'set error', true));
        const inputElement = component.find('input');
        setImmediate(() => {
            component.update();
            expect(component.find('.invalid-feedback').text()).toBe('set error');
            expect(component.find('input').hasClass('is-invalid')).toBeTruthy();
            done();
        })
    });
});
