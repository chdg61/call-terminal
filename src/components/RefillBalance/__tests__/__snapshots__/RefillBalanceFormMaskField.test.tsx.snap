// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefillBalanceFormMaskField default render 1`] = `"<label for=\\"phone-test\\">Номер телефона</label><input class=\\"form-control\\" id=\\"phone-test\\" placeholder=\\"Введите номер телефона\\" name=\\"phone-test\\" value=\\"\\"/>"`;
